README.md
# hyperproxy [![Build Status](https://secure.travis-ci.org/brianloveswords/hyperproxy.png?branch=master)](http://travis-ci.org/brianloveswords/hyperproxy)

Reverse proxy with advanced routing capabilities

Uses <code>[urlglob](https://github.com/brianloveswords/urlglob)</code> for route matching

**NOTE** This project is still early days. Use it on personal servers, side projects and other applications that aren't super critical. Please file bugs to help me track down issues!

For some battle-hardened alternatives, check out nodejitsu's <code>[node-http-proxy](https://github.com/nodejitsu/node-http-proxy)</code> and dotCloud's <code>[hipache](https://github.com/dotcloud/hipache)</code>

## Install

```bash
$ npm install hyperproxy
```

## CLI

Checkout <a href="//github.com/brianloveswords/hyperproxy-cli"><code>hyperproxy-cli</code></a>

## Example
```js
const Hyperproxy = require('hyperproxy')

const proxy = new Hyperproxy({
  servers: [
    // routes are tested in order, so exact matches should come first
    // and less specific routes afterwards

    // exact matches, routes to local ports
    [ 'pi.example.org', ':3141'  ],

    // HTTPS servers require a more explicit definition
    { pattern: 'tau.example.org',
      endpoint: ':1618',
      https: {
        key: fs.readFileSync('/path/to/tau-key.pem'),
        cert: fs.readFileSync('/path/to/tau-cert.pem'),
      }
    },

    // `forceSecure` will force a 301 redirect to the HTTPS server when
    // the HTTP server is hit. You should also consider adding a
    // `Strict-Transport-Security` header.
    { pattern: 'euler.example.org',
      endpoint: ':2718',
      forceSecure: true,
      https: {
        key: fs.readFileSync('/path/to/euler-key.pem'),
        cert: fs.readFileSync('/path/to/euler-cert.pem'),
      }
    },

    // route to an external domain
    [ 'google.example.org', 'google:80' ],

    // matches
    // - 'images.example.org'
    // - 'stuff.user.example.org'
    // - '⚡.example.org'
    [ '*.example.org', '/tmp/any-subdomain.socket' ],

    // url matching!
    [ 'example.org', [
      // matches all sub paths, e.g. /static/a/b/
      ['/static/*', '/tmp/static.socket' ],

      // use '*?' to match one level deep, e.g '/js/x.js',
      // but not '/js/sub/x.js'
      ['/js/*?', '/tmp/javascript.socket' ],

      // matches '/v2/x/y/z.json', '/v22.73/stuff.json'
      ['/api/*.json', '/tmp/json-api.socket' ],

      // matches '/v1/x/y/z.xml', '/vπ/stuff.xml'
      ['/api/*.xml', '/tmp/xml-api.socket' ],

      // handle anything that falls through.
      [ '*', '/tmp/default.socket' ],
    ]],

    // this is exactly the same as above, only it uses the explicit
    // style of configuration instead of the implicit style.
    { pattern: 'example.org',
      routes: [
        { pattern: '/static/*', endpoint: '/tmp/static.socket' },
        { pattern: '/js/*?', endpoint: '/tmp/javascript.socket' },
        { pattern: '/api/*.json', endpoint: '/tmp/json-api.socket' },
        { pattern: '/api/*.xml', endpoint: '/tmp/xml-api.socket' },
        { pattern '*', endpoint: '/tmp/default.socket' },
      ]
    },
  ]
})

const server = proxy
  .createServer(function(req, res, proxyRoute){
    // callback is optional, if nothing is passed proxying will
    // happen as normal. when a callback is passed, calling `proxyRoute`
    // will continue on to the proxy routing process.
    res.headers['x-proxy'] = 'hyperproxy'
    proxyRoute() // continue to proxy routing
  }).listen(80)


// `createServer` only makes the HTTP server, if you want to also
// make the HTTPS server, you have to call `createSecureServer`.
const secureServer = proxy
  .createSecureServer()
  .listen(443)
```
### A Note About HTTPS

Hyperproxy uses [Server Name Indication](https://en.wikipedia.org/wiki/Server_Name_Indication) to choose different certificates based on hostname. While SNI has wide support at this point, [not all platforms support it](https://en.wikipedia.org/wiki/Server_Name_Indication#No_support), notably Windows XP. Microsoft is [ending support for Windows XP](http://windows.microsoft.com/en-us/windows/end-support-help), on 2014-04-08 so it is expected that the userbase with SNI capability will grow as the last holdouts on XP upgrade their systems.


### Error Handling

There are number of things that can go wrong when trying to proxy a request: it could not match any routes, the socket could be missing or dead, the remote host could be down or something else completely unexpected.

When this happens, the default method of handling it will be to return a `HTTP 502: Bad Gateway` to the requesting client. However, if you attach an event handler to any one of the events below, the default behavior will be bypassed (though you will still be able to invoke it from the event handler).

All event handlers **other than `proxyError`**  get four arguments: `(err, req, res, defaultHandler)`. `proxyError` only receives the error.

`defaultHandler` is a reference to the default error handler. For example, if you wanted to log all unknown errors and passthrough to the default handler, you could do the following:


#### <code>proxyError</code>
Gets emitted on any error, in addition to the more specific events below. This can be useful for logging.

#### <code>proxyMiss</code>
A route couldn't be found

#### <code>missingSocketFile</code>
Couldn't find the socket file associated with the route

#### <code>hostNotFound</code>
Couldn't find the host associated with the route

#### <code>unknownError</code>
Catch-all for any other problems that occur when trying to attach an endpoint to a request.


```js
server.on('unknownError', function(err, req, res, defaultHandler){
  // use your log handler
  logger.log('unknown proxy error', err)
  defaultHandler()
})
```


## Current Limitations
These may be implemented as plugins later on.

* `HTTP` and `HTTPS` only, no websocket support yet.
* No load balancing, caching or any "advanced" features.

## License

MIT

```
Copyright (c) 2013 Brian J. Brennan

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

CHANGELOG.md
# hyperproxy changelog

## v0.6.0
* [breaking] rename `secureOnly` to `forceSecure`, which is a better name for what the option actually does.

## v0.5.3
* Add `secureOnly` feature, which redirects from an insecure server to the secure one.

## v0.5.0
* Add support for HTTPS, serving multiple certificates with SNI
* Add “explicit” style of declaring servers, using objects instead of arrays.

test/server.js
const fs = require('fs')
const http = require('http')
const concat = require('concat-stream')

module.exports = makeServer

function makeServer(socketOrPort, opts) {
  opts = opts || {}
  socketOrPort = cleanupSocket(socketOrPort)

  console.log('starting test server on', socketOrPort)

  var listeningPort
  const server = http.createServer(function (req, res) {
    const headers = req.headers
    const host = headers.host
    const path = req.url
    const method = req.method

    req.pipe(concat(function (data) {
      res.writeHead(opts.statusCode || 200, opts.headers)
      res.write(JSON.stringify({
        requestHeaders: headers,
        socketPath: socketOrPort,
        port: listeningPort,
        host: host,
        path: path,
        method: method,
        body: opts.body,
        postData: data.toString(),
      }, null, '  '))

      res.end()
    }))

  }).listen(socketOrPort)

  server.on('listening', function () {
    listeningPort = this.address().port
  })

  server.unref && server.unref()
  server.socket = socketOrPort
  return server
}

function cleanupSocket(socket) {
  if (socket.indexOf(':') == 0)
    return socket.slice(1)

  try { fs.unlinkSync(socket) }
  catch (e) { if (e.code != 'ENOENT') throw e}
  finally { return socket }
}


test/request.js
const path = require('path')
const http = require('http')
const https = require('https')
const concat = require('concat-stream')

module.exports = function (opts, callback) {
  opts = opts || {}
  const requestMethod = (opts.https == true)
    ? https.request.bind(https)
    : http.request.bind(http)

  const request = requestMethod({
    rejectUnauthorized: false,
    socketPath: opts.socketPath,
    port: opts.port,
    path: opts.path,
    method: opts.method,
    hostname: opts.hostname,
    headers: {
      "host": opts.hostname,
      "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:28.0) Gecko/20100101 Firefox/28.0",
    },
  }, function (res) {
    res.pipe(concat(function (body) {
      if (opts.json === false)
        return callback(body)
      const response = JSON.parse(body.toString())
      const requestHeaders = response.requestHeaders
      const responseHeaders = res.headers
      return callback(response, requestHeaders, responseHeaders, res.statusCode)
    }))
  })
  request.end(opts.postData || Buffer(''))
  return request
}


test/servers.js
const fs = require('fs')
const path = require('path')
const testServer = require('./server')
const localSocket = require('./localsocket')
const xtend = require('xtend')

module.exports = {
  gnarly: new Server('gnarly'),
  tubular: new Server('tubular'),
  wayCool: new Server('wayCool'),
  awesome: new Server('awesome'),
  groovy: new Server('groovy'),
  mondo: new Server('mondo'),
  outrageous: new Server('outrageous'),
  funky: new Server('funky'),
}

function Server(name) {
  this.name = name
  this.socket = localSocket(name + '.sock')
}

Server.prototype.tlsOptions = function () {
  const cert = this.name + '.localhost.crt'
  const opts =  {
    key: fs.readFileSync(path.join(__dirname, 'cert', 'ia.key')),
    cert: fs.readFileSync(path.join(__dirname, 'cert', cert)),
  }
  console.dir(opts)
  return opts
}

Server.prototype.start = function (opts) {
  this.server = testServer(this.socket, this.options(opts))
  return this.socket
}

Server.prototype.options = function (opts) {
  return xtend({
    statusCode: '200',
    headers: {
      'x-server-name': this.name,
      'x-server-socket': this.socket,
    },
    body: this.message()
  }, opts)
}

Server.prototype.message = function () {
  return 'totally ' + this.name + ' bro.'
}


urlglob.js
const LRU = require('lru-cache')
const cache = LRU(500)

const sep = '%'+Date.now()+'%'

module.exports = urlglob

function urlglob(pattern, test) {
  var regex;
  const regexFromCache = cache.get(pattern)
  if (regexFromCache)
    regex = regexFromCache
  else {
    regex = convert(pattern)
    cache.set(pattern, regex)
  }
  if (!test) return regex

  const cacheKey = [pattern, sep ,test].join('')
  const matchFromCache = cache.get(cacheKey)

  if (typeof matchFromCache != 'undefined')
    return matchFromCache

  const match = regex.test(test)
  cache.set(cacheKey, match)
  return match
}

function convert(pattern) {
  const escaped = pattern.replace(/([()[{+.$^\\|?])/g, '\\$1');
  const placeholder = '%STAR%' + Date.now() + '%'
  const catchAll = '.*?'
  const regexp = '^' + escaped
    .replace(/\\[*]/g, placeholder)
    .replace(/\*/, catchAll)
    .replace(RegExp(placeholder, 'g'), '*')
    + '\\/?$'
  return new RegExp(regexp)
}

urlglob.inCache = function inCache(pattern) {
  return cache.get(pattern)
}

urlglob.cache = cache
urlglob.convert = convert


index.js
const pkg = require('./package')
const fs = require('fs')
const path = require('path')
const find = require('./find')
const http = require('http')
const https = require('https')
const crypto = require('crypto')
const urlglob = require('urlglob')

module.exports = Hyperproxy

function Hyperproxy(opts, callback) {
  if (!(this instanceof Hyperproxy))
    return new Hyperproxy(opts, callback)

  opts = Hyperproxy.normalizeOptions(opts)
  this.https = opts.https
  this.servers = opts.servers
  this.agent = opts.agent
    ? opts.agent
    : Hyperproxy.defaultAgent()

  if (callback)
    return this.createServer(callback)
}

Hyperproxy.VERSION = pkg.version

Hyperproxy.prototype = {
  changeConfig: function changeConfig(opts) {
    opts = Hyperproxy.normalizeOptions(opts)
    this.https = opts.https
    this.servers = opts.servers
  },
  createServer: function createServer(callback) {
    callback = callback || Hyperproxy.connectionNoop
    const server = http.createServer()
    const requestHandler = makeRequestHandler({
      secure: false,
      server: server,
      callback: callback,
      context: this
    })
    server.on('request', requestHandler)

    return server
  },
  createSecureServer: function createSecureServer(callback) {
    callback = callback || Hyperproxy.connectionNoop
    const servers = this.servers
    if (!this.https)
      throw new Error('At least one endpoint must have https options to create a secure server')

    const context = {}
    const secureServer = https.createServer({
      key: this.https.key,
      cert: this.https.cert,
      pfx: this.https.pfx,
      SNICallback: function (servername) {
        const cachedContext = context[servername]
        if (cachedContext)
          return cachedContext

        const secureServers = servers.filter(function (server) {
          return server.https
        })

        const server = find(secureServers, function (server) {
          const hostPattern = server.pattern
          return urlglob(hostPattern, servername)
        })

        if (!server)
          return false

        const secureContext =
          crypto.createCredentials(server.https).context

        secureServer.emit('certificateNegotiation', servername, server)
        context[servername] = secureContext
        return secureContext
      }
    })

    const secureRequestHandler =
      makeRequestHandler({
        secure: true,
        server: secureServer,
        callback: callback,
        context: this
      })

    secureServer.on('request', secureRequestHandler)
    return secureServer
  },
  pickServer: function (host) {
    return find(this.servers, function (server) {
      const hostPattern = server.pattern
      return urlglob(hostPattern, host)
    })
  },
}

function makeRequestHandler(opts) {
  const server = opts.server
  const callback = opts.callback
  const secureRequest = opts.secure
  const ctx = opts.context || opts.ctx
  return function(clientReq, clientRes) {
    function proxyMiss(server, req, res) {
      const fullUrl = req.headers.host + req.url
      const error = new Error('Could not find endpoint for "' + fullUrl + '"')
      error.code = 'proxyMiss'
      const handler = Hyperproxy.errorHandler(error)
      return handler({
        error: error,
        server: server,
        req: clientReq,
        res: clientRes
      })
    }

    function makeProxyRequest() {
      if (!clientReq.headers.host)
        return proxyMiss(server, clientReq, clientRes)

      const host = clientReq.headers.host.split(':')[0]
      const serverConfig = ctx.pickServer(host)
      const requestOpts = Hyperproxy.createRequestOpts({
        servers: ctx.servers,
        headers: clientReq.headers,
        host: host,
        path: clientReq.url,
        method: clientReq.method,
      })

      if (!requestOpts)
        return proxyMiss(server, clientReq, clientRes)

      if (serverConfig.forceSecure && !secureRequest)
        return httpsRedirect(clientReq, clientRes)

      const proxyReq = http.request(requestOpts, endpointResponse)
      function endpointResponse(proxyRes) {
        clientRes.writeHead(proxyRes.statusCode, proxyRes.headers)
        proxyRes.pipe(clientRes)
      }; clientReq.pipe(proxyReq)

      proxyReq.on('error', function (err) {
        const handler = Hyperproxy.errorHandler(err)
        return handler({
          error: err,
          server: server,
          req: clientReq,
          res: clientRes
        })
      })
    }
    callback.apply(ctx, [
      clientReq,
      clientRes,
      makeProxyRequest
    ])
  }
}

function httpsRedirect(req, res) {
  const url = 'https://' + req.headers.host +  req.url
  res.writeHead(301, {Location: url})
  res.end()
}

Hyperproxy.normalizeOptions = function normalizeOptions(opts) {
  opts.servers = opts.servers.map(function (server) {
    // array style: [pattern, endpoint]
    if (Array.isArray(server))
      server = { pattern: server[0], endpoint: server[1] }

    if (server.https)
      opts.https = server.https = fixHttpsOptions(server.https)

    var routes = server.endpoint || server.upstream || server.routes

    if (Array.isArray(routes)) {
      delete server.endpoint
      delete server.upstream
      server.routes = routes.map(function (route) {
        if (Array.isArray(route))
          return {
            pattern: route[0],
            endpoint: route[1],
            forceSecure: server.forceSecure,
          }
        return route
      })
    }
    return server
  })
  return opts
}

function readFile(file) {
  return fs.readFileSync(path.resolve(file))
}
function fixHttpsOptions(https) {
  if (https.key && !Buffer.isBuffer(https.key))
    https.key = readFile(https.key)
  if (https.cert && !Buffer.isBuffer(https.cert))
    https.cert = readFile(https.cert)
  if (https.pfx && !Buffer.isBuffer(https.pfx))
    https.pfx = readFile(https.pfx)
  return https
}

Hyperproxy.createRequestOpts = function createRequestOpts(opts) {
  const servers = opts.servers
  const path = opts.path

  const server = find(servers, function (server) {
    const hostPattern = server.pattern
    return urlglob(hostPattern, opts.host)
  })

  if (!server)
    return false

  var endpoint = server.endpoint || server.upstream || server.routes

  if (Array.isArray(endpoint)) {
    const routes = endpoint
    const found = find(routes, function (route) {
      return urlglob(route.pattern, path)
    })
    if (found)
      endpoint = found.endpoint
  }

  if (!endpoint)
    return false

  return Hyperproxy.finishRequestOpts(endpoint, {
    path: opts.path,
    headers: opts.headers,
    method: opts.method,
  })
}

Hyperproxy.finishRequestOpts =
  function finishRequestOpts(endpoint, opts) {
    if (Hyperproxy.isSocket(endpoint)) {
      opts.socketPath = endpoint
      return opts
    }

    if (Hyperproxy.isPort(endpoint)) {
      opts.port = (''+endpoint).split(':')[1]
      return opts
    }

    const parts = endpoint.split(':')
    const hostname = parts[0]
    const port = parts[1]

    opts.headers.host = hostname
    opts.hostname = hostname
    opts.port = port

    return opts
  }

Hyperproxy.errorHandler = function errorHandler(error) {
  return ((Hyperproxy.errorHandlers[error.code]  ||
          Hyperproxy.errorHandlers.unknown)
          .bind(Hyperproxy.errorHandlers))
}

Hyperproxy.errorHandlers = {
  ENOENT: function (opts) {
    return Hyperproxy.errorEvent('missingSocketFile', opts)
  },
  ENOTFOUND: function (opts) {
    return Hyperproxy.errorEvent('hostNotFound', opts)
  },
  proxyMiss: function (opts) {
    return Hyperproxy.errorEvent('proxyMiss', opts)
  },
  unknown: function (opts) {
    return Hyperproxy.errorEvent('unknownError', opts)
  },
}

Hyperproxy.errorEvent = function (event, opts) {
  const server = opts.server
  const handleBadGateway =
    Hyperproxy.badGateway.bind(Hyperproxy, opts)
  server.emit('proxyError', opts.error)
  if (!server.listeners(event).length)
    return handleBadGateway()

  server.emit(event, opts.error, opts.req, opts.res, handleBadGateway)
}

Hyperproxy.connectionNoop = function connectionNoop(_, _, done) {
  return done()
}


Hyperproxy.isSocket = function isSocket(input) {
  return (input.indexOf(':') == -1 ||
          typeof input == 'number' ||
          !isNaN(Number(input)))
}

Hyperproxy.isPort = function isPort(string) {
  return string.indexOf(':') == 0
}

Hyperproxy.defaultAgent = function defaultAgent() {
  const agent = new http.Agent()
  agent.maxSockets = Infinity
  return agent;
}

Hyperproxy.badGateway = function badGateway(opts) {
  const res = opts.res
  res.writeHead(502, 'Bad Gateway')
  res.write('Bad Gateway')
  return res.end()
}

Hyperproxy.error = function error(msg, obj) {
  if (typeof msg == 'object')
    msg = msg.message, obj = msg
  const err = new Error(msg)
  err.code = obj.code
  err.original = obj.original
}
